### PR TITLE
feat(kit): `InputPhone` uses `Maskito` instead of legacy `text-mask`

### LINKED ISSUE #3422: 🐞 - Incorrect paste behavior with selected text in tui-input-phone
### Which @taiga-ui/* package(s) are the source of the bug?

kit

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-6aeked?file=src/app/app.component.html

### Is this issue blocking you?

Non-Blocking

### Description

**Steps to reproduce:**

1. Input number to tui-input-phone
2. Select some piece of text
3. Paste some number via Ctrl-V/Cmd-v

**Expected behaviour:**
Only selected text will be changed

**Real behaviour**
Whole text will be replaced by pasted number

https://user-images.githubusercontent.com/23421487/212280827-7db1048b-47c6-4591-9455-a668b385100a.mov

### Angular version

12

### Taiga UI version

3.15.0

### Which browsers have you used?

- [X] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Edge

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

### LINKED ISSUE #3451: 🐞 - Invalid filler for `phoneNumber`
### Which @taiga-ui/* package(s) are the source of the bug?

kit

### Please provide a link to a minimal reproduction of the bug

https://taiga-ui.dev/components/input-phone/API?tuiTextfieldCustomContent=tuiIconMastercardMono&phoneMaskAfterCountryCode=(%23%23%23)%20%23%23%23-%23%23-%23%23&tuiTextfieldCleaner=true&focusable=true&tuiTextfieldPrefix=&tuiTextfieldPostfix=&tuiTextfieldFiller=57567567&tuiTextfieldSize=l&search=q3e

### Is this issue blocking you?

Non-Blocking

### Description

image

### Angular version

latest

### Taiga UI version

latest

### Which browsers have you used?

- [X] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

### PR DESCRIPTION
**Pull request by nsbarsukov**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [X] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

## What is the current behavior?
Partially solves https://github.com/Tinkoff/taiga-ui/issues/120
Closes #3422
Closes #3451

### CODE REVIEW

File: projects/core/utils/format/format-phone.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import {CHAR_PLUS} from '@taiga-ui/cdk';
 
 /**
+ * TODO: Guys, should we mark this function as deprecated?
```
Comment (waterplea) [inline projects/core/utils/format/format-phone.ts:4]: Sure, just make sure to add example on how to use new maskitoTransform in jsdocs

File: projects/kit/components/input-phone/input-phone.component.ts [hunk 2]
```diff
@@ -151,10 +133,18 @@ export class TuiInputPhoneComponent
...
-            : this.search || '';
+    get nativeValue(): string {
+        return this.nativeFocusableElement
+            ? this.nativeFocusableElement.value
+            : maskitoTransform(this.value, this.maskOptions);
+    }
+
+    set nativeValue(value: string) {
+        if (!this.nativeFocusableElement) {
+            return;
```
Comment (waterplea) [inline projects/kit/components/input-phone/input-phone.component.ts:144]: nit: with one line we do not do early returns

File: projects/kit/components/input-phone/input-phone.component.ts [hunk 3]
```diff
@@ -166,56 +156,41 @@ export class TuiInputPhoneComponent
...
-        if (
-            this.computedValue === this.countryCode ||
-            (this.search !== null &&
-                Number.isNaN(
-                    parseInt(this.search.replace(TUI_MASK_SYMBOLS_REGEXP, ''), 10),
-                ))
-        ) {
-            this.updateSearch('');
+        if (!active && !this.allowText && this.nativeFocusableElement) {
+            this.nativeValue = this.nativeValue.replace(/-$/, '');
```
Comment (waterplea) [inline projects/kit/components/input-phone/input-phone.component.ts:184]: Why only hyphen is removed? Can it be space?

File: projects/kit/components/input-phone/input-phone.component.ts [hunk 4]
```diff
@@ -166,56 +156,41 @@ export class TuiInputPhoneComponent
...
 
-    onBackspace(event: Event): void {
-        const target = event.target as HTMLInputElement;
-
-        if (
-            (target.selectionStart || 0) <= this.nonRemovableLength &&
-            target.selectionStart === target.selectionEnd
-        ) {
-            event.preventDefault();
+        if (this.nativeValue === this.nonRemovablePrefix || this.isTextValue) {
```
Comment (waterplea) [inline projects/kit/components/input-phone/input-phone.component.ts:187]: Why are we clearing text value?

File: projects/kit/components/input-phone/input-phone.component.ts [hunk 5]
```diff
@@ -278,6 +259,32 @@ export class TuiInputPhoneComponent
...
+        return allowText
+            ? {
+                  mask: ({value}) =>
+                      isText(value) && value !== '+'
+                          ? (MASKITO_DEFAULT_OPTIONS.mask as RegExp)
+                          : mask,
+              }
+            : {
+                  mask,
+                  preprocessor: tuiCreateCompletePhoneInsertionPreprocessor(
```
Comment (waterplea) [inline projects/kit/components/input-phone/input-phone.component.ts:280]: We do not need those in the `allowText` version?

File: projects/kit/components/input-phone/input-phone.component.ts (hunk 4 above)
Comment (nsbarsukov) [reply projects/kit/components/input-phone/input-phone.component.ts:187]: It is already existing behaviour for `main` branch:
https://taiga-ui.dev/components/input-phone#autocomplete


https://github.com/Tinkoff/taiga-ui/assets/35179038/5354ab7b-3b1f-4c46-b012-bc80d6ae232d

Comment (nsbarsukov) [reply projects/kit/components/input-phone/input-phone.component.ts:187]: I think that it is kind of feature 🤔

File: projects/kit/components/input-phone/input-phone.component.ts (hunk 3 above)
Comment (nsbarsukov) [reply projects/kit/components/input-phone/input-phone.component.ts:184]: I can change to this way (not digit):

```suggestion
            this.nativeValue = this.nativeValue.replace(/\D$/, '');
```

What do you think about it ?

File: projects/kit/components/input-phone/input-phone.component.ts (hunk 2 above)
Comment (nsbarsukov) [reply projects/kit/components/input-phone/input-phone.component.ts]: fixed

File: projects/kit/components/input-phone/input-phone.component.ts (hunk 5 above)
Comment (nsbarsukov) [reply projects/kit/components/input-phone/input-phone.component.ts]: fixed

File: projects/kit/components/input-phone/input-phone.component.ts (hunk 4 above)
Comment (waterplea) [reply projects/kit/components/input-phone/input-phone.component.ts:185]: I guess it is so only valid "named" phone numbers are there in the control.

Comment (waterplea) [reply projects/kit/components/input-phone/input-phone.component.ts:185]: I guess it is so only valid "named" phone numbers are there in the control.

File: projects/kit/components/input-phone/input-phone.component.ts (hunk 3 above)
Comment (waterplea) [reply projects/kit/components/input-phone/input-phone.component.ts]: LGTM

File: projects/core/utils/format/format-phone.ts [hunk 6]
```diff
@@ -1,6 +1,16 @@
 import {CHAR_PLUS} from '@taiga-ui/cdk';
 
 /**
+ * TODO: delete in v4.0
+ * @deprecated Use {@link https://tinkoff.github.io/maskito/core-concepts/transformer maskitoTransform} instead<br/>
+ * @example
+ * import {maskitoTransform} from '@maskito/core';
+ *
+ * const mask = {
+ *     mask: ['+', '1', ' ', /\d/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/],
```
Comment (waterplea) [inline projects/core/utils/format/format-phone.ts:10]: I wonder if we need a public API instrument to turn ##-### mask representation to `MaskitoOptions` 🤔

**Review by waterplea — APPROVED** [review]

**Review by vladimirpotekhin — APPROVED** [review]

**Pull request merged**